tower-http/src/validate_request.rs

```diff
@@ -0,0 +1,551 @@
+{
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.project().kind.project() {
+            KindProj::Future { future } => future.poll(cx),
+            KindProj::Error { response } => {
+                let response = response.take().expect("future polled after completion");
+                Poll::Ready(Ok(response))
+            }
+        }
+    }
+}
+
+/// Trait for validating requests.
+pub trait ValidateRequest<B> {
+    /// The body type used for responses to unvalidated requests.
+    type ResponseBody;
+
+    /// Validate the request.
+    ///
+    /// If `Ok(())` is returned then the request is allowed through, otherwise not.
+    fn validate(&mut self, request: &mut Request<B>) -> Result<(), Response<Self::ResponseBody>>;
+}
+
+impl<B, F, ResBody> ValidateRequest<B> for F
+where
+    F: FnMut(&mut Request<B>) -> Result<(), Response<ResBody>>,
+{
+    type ResponseBody = ResBody;
+
+    fn validate(&mut self, request: &mut Request<B>) -> Result<(), Response<Self::ResponseBody>> {
+        self(request)
+    }
+}
+
+/// Type that performs validation of the Accept header.
+pub struct AcceptHeader<ResBody> {
+    header_value: Arc<Mime>,
+    _ty: PhantomData<fn() -> ResBody>,
+}
+
+impl<ResBody> AcceptHeader<ResBody> {
+    /// Create a new `AcceptHeader`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `header_value` is not in the form: `type/subtype`, such as `application/json`
+    fn new(header_value: &str) -> Self
+    where
+        ResBody: Body + Default,
+    {
+        Self {
+            header_value: Arc::new(
+                header_value
+                    .parse::<Mime>()
+                    .expect("value is not a valid header value"),
+            ),
+            _ty: PhantomData,
+        }
+    }
+}
+
+impl<ResBody> Clone for AcceptHeader<ResBody> {
+    fn clone(&self) -> Self {
+        Self {
+            header_value: self.header_value.clone(),
+            _ty: PhantomData,
+        }
+    }
+}
+
+impl<ResBody> fmt::Debug for AcceptHeader<ResBody> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("AcceptHeader")
+            .field("header_value", &self.header_value)
+            .finish()
+    }
+}
+
+impl<B, ResBody> ValidateRequest<B> for AcceptHeader<ResBody>
+where
+    ResBody: Body + Default,
+{
+    type ResponseBody = ResBody;
+
+    fn validate(&mut self, req: &mut Request<B>) -> Result<(), Response<Self::ResponseBody>> {
+        if !req.headers().contains_key(header::ACCEPT) {
+            return Ok(());
+        }
+        if req
+            .headers()
+            .get_all(header::ACCEPT)
+            .into_iter()
+            .flat_map(|header| {
+                header
+                    .to_str()
+                    .ok()
+                    .into_iter()
+                    .flat_map(|s| s.split(",").map(|typ| typ.trim()))
```

Comment:
Worth noting that this is incorrect for the same reason as this issue https://github.com/http-rs/http-types/issues/349

Reply:
Oh thanks for catching that! Would you mind opening an issue?

Reply:
I might not be the best at doing this — I don't really know Rust and just stumbled on this when giving a colleague advice on how to port something from a Node project to their Rust project. But I can if you would prefer an issue from somebody who doesn't actually directly use your project or know Rust :)

Reply:
Alright. I've opened https://github.com/tower-rs/tower-http/issues/291. Thanks for reporting this.